streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStore.java

```diff
@@ -0,0 +1,792 @@
+                    // tombstone's timestamp as both validFrom and validTo timestamps for the segment
+                    segment.put(key, segmentValueSchema.newSegmentValueWithRecord(null, timestamp, timestamp).serialize());
+                } else {
+                    // insert as latest, since foundTs = sentinel means nothing later exists
+                    if (segmentValueSchema.getNextTimestamp(segmentValue) == timestamp) {
+                        // next timestamp equal to put() timestamp already represents a tombstone,
+                        // so no additional insertion is needed in this case
+                        return;
+                    }
+                    final SegmentValue sv = segmentValueSchema.deserialize(segmentValue);
+                    sv.insertAsLatest(
+                        segmentValueSchema.getNextTimestamp(segmentValue),
+                        timestamp,
+                        null
+                    );
+                    segment.put(key, sv.serialize());
+                }
+            }
+        } else {
+            // insert into segment corresponding to foundTs, as foundTs represents the validTo
+            // timestamp of the current put.
+            // the new record is either the earliest or the latest in this segment, depending on the
+            // circumstances of the fall-through. (it cannot belong in the middle because otherwise
+            // putSegments() above would have identified a segment for which
+            // minTimestamp <= timestamp < nextTimestamp, and putSegments would've completed the
+            // put procedure without reaching this fall-through case.)
+            final T segment = versionedStoreClient.getOrCreateSegmentIfLive(
+                versionedStoreClient.segmentIdForTimestamp(foundTs), context, observedStreamTime);
+            if (segment == null) {
+                if (expiredRecordSensor.isPresent()) {
+                    expiredRecordSensor.get().record(1.0d, ProcessorContextUtils.currentSystemTime(context));
+                    LOG.warn("Skipping record for expired put.");
+                }
+                return;
+            }
+
+            final byte[] segmentValue = segment.get(key);
+            if (segmentValue == null) {
+                if (value != null) {
+                    segment.put(
+                        key,
+                        segmentValueSchema.newSegmentValueWithRecord(value, timestamp, foundTs).serialize()
+                    );
+                } else {
+                    segment.put(
+                        key,
+                        segmentValueSchema.newSegmentValueWithRecord(null, timestamp, foundTs).serialize()
+                    );
+                }
+            } else {
+                final long foundNextTs = segmentValueSchema.getNextTimestamp(segmentValue);
+                if (foundNextTs <= timestamp) {
```

Comment:
I am just wondering if this case is actually possible? Above you just say vaguely `depending on the circumstances of the fall-through` -- can you elaborate?

The only case for "latest" is do understand would be the case when we actually insert into the "latest value store" instead of a segment. If we really insert into a segment, it seems we would always insert as earliest?

Reply:
Nope, latest is possible for a segment as well. It has to do with the "gaps" which may be formed from degenerate segments. Here's a example:
* segment interval is 100
* put tombstone with `ts=50`, so there is a (degenerate) segment with `minTs=nextTs=50` (containing a single tombstone).
* put record with `ts=150`, so it goes to the latest value store
* put record with `ts=90`, so now there is a new segment with `minTs=90`, `nextTs=150`, containing the new record.
* put record with `ts=80`: this put operation will "fall-through" with `foundTs=90`, so the segment it finds will be the one which contains a tombstone (with `minTs=nextTs=50`), so `foundNextTs=50` which is less than the insertion timestamp (80)`.

Do you think this is worth putting into a comment in the code? It's a very long explanation, and requires knowledge of "degenerate segments" which this class currently does not require. I'll leave it out for now.